make winning possible

DerpyNewbie opened the pull request.

DerpyNewbie committed:
rebase
Add checks for game win/draw

Fix highlighting (again)
wait... i can't select which file to commit?
TicTacToeBlock- Add SetType(type)
TicTacToeBlockGrid- Add ResetGame(), OnNextTurn()
maybe use #OnNextTurn() to handle AI stuff?

Fix highlight issue
if statement should be checking BlockType EQUALS 0 :O

Revert "Fix highlight issue"
This reverts commit 0fe7a3d85e0595bc825e6051d8c8ad167ddd3300.

I-K0haku-I commented on Source/TicTacToe/TicTacToeBlockGrid.h:
```diff
@@ -26,9 +27,15 @@ class ATicTacToeBlockGrid : public AActor
 	int32 Score;
 
-	/** Current turn role by int8. 1 to circle, -1 to cross.*/
+	/** Current turn role by int. 1 to circle, -1 to cross. */
```
test

I-K0haku-I commented on Source/TicTacToe/TicTacToeBlock.cpp:
```diff
@@ -105,7 +98,19 @@ void ATicTacToeBlock::Highlight(bool bOn)
 	else
-	{
-		BlockMesh->SetMaterial(0, BaseMaterial);
+	{	// Using GetMaterialByType for fail safe
```
Unclear what you mean here.

DerpyNewbie replied:
Setting `BaseMaterial` directly made "confused block" like this, when i failed to fix "if returning check" on the `line:91`.
`(BlockType = 1 (Circle), Material = BaseMaterial (should be OrangeMaterial))`

So, i decided to instead of changing to `BaseMaterial` directly, changing from `GetMaterialByType()` by current `BlockType` to make sure the unhighlighted blocks are having corrected materials.

TL:DR
I changed BaseMaterial to GetMaterialByType so when accidentally called this method, it won't reset materials. (that's what i mean by fail safe)

I-K0haku-I replied:
Unnecessary information then. Can be removed.

DerpyNewbie replied:
Removed unnecessary comment on latest commit `5e633dc351346d17ab6b0cb65152d6ad49530768`

I-K0haku-I commented on Source/TicTacToe/TicTacToeBlock.cpp:
```diff
@@ -105,7 +98,19 @@ void ATicTacToeBlock::Highlight(bool bOn)
+{
+	SetMaterialByType(_type);
+	BlockType = _type;
+}
```
Why do we need SetMaterialByType, just do it like BlockMesh->SetMaterial(0, GetMaterialByType(BlockType)); above.
Or use a method for every assignment. You could just do SetType(BlockType) above instead.

Either way, you are repeating yourself here.

DerpyNewbie replied:
I couldn't figure out how to set materials on `TicTacToeBlockGrid`. 
(i tried to do `executedBlock->BlockMesh->SetMaterial()` but didn't work at least on my visual studio intellisense.)

So, i made `SetMaterialByType()` to set material on `TicTacToeBlockGrid`.
but i actually didn't like calling setter 2 times in `TicTacToeBlockGrid`.

so i decided to add `SetType()` to set type with one method.
since it's not used by `TicTacToeBlockGrid`, `SetMaterialByType()` should be private/protected method or deleted.

TL:DR
`SetMaterialByType()` should be deprecated or private/protected method, since i made `SetType()` for setting types.

I-K0haku-I replied:
Create a ResetMaterial method and use SetType for actually setting it. SetType might not even be necessary if we move most of the calls into Block.

DerpyNewbie replied:
Changed `SetMaterialByType()`'s access specifier to private on latest commit `5e633dc351346d17ab6b0cb65152d6ad49530768`
(still used in `ATicTacToeBlock` class. i feel like it's little bit more easy to read.)

I-K0haku-I commented on Source/TicTacToe/TicTacToeBlockGrid.cpp:
```diff
@@ -74,4 +77,127 @@ void ATicTacToeBlockGrid::ChangeTurn()
+		ChangeTurn();
+		OnNextTurn();
+	}
+}
```
Remove executedBlock from this and move the logic that references it into the actual Block. Everything else you can leave because it doesn't need the block.

You could just use the Grid reference inside the Block for all of this too.

DerpyNewbie replied:
I'm still not sure about this.
i think those stuff should be in `TicTacToeBlockGrid`.
since they are deciding what to do by stuff what `ATicTacToeBlockGrid` have.

I-K0haku-I replied:
@DerpyNewbie 
That's fine, you can keep it like that then. Probably just a taste thing.

But the main thing is how confusing the code latter part of the code is because you are doing so much at once. The worst offender is ```checkWin``` because it returns both block type and the game state even though you catch it as ```winState```. That's probably why the conditional blocks after that look so confusing.

I-K0haku-I replied:
Another thing: The EWinState you receive is ONLY used in this HandleTurn method and nowhere else, correct me if I'm wrong.
You could call the method in the conditional block inside CheckWinState instead of setting the Enums and it would be the same.

Furthermore, I should move the 'can win' method from my AI to the static class. This way, we can just call it in CheckWinState. So no need to change anything right now. (Same with GetIndex)

I-K0haku-I commented on Source/TicTacToe/TicTacToeBlockGrid.cpp:
```diff
@@ -74,4 +77,127 @@ void ATicTacToeBlockGrid::ChangeTurn()
+int ATicTacToeBlockGrid::GetPosition(ATicTacToeBlock* block_pos)
+{
+	return BlockArray.Find(block_pos);
+}
```
Rename these into GetIndex. Position is confusing.

DerpyNewbie replied:
Renamed to `GetIndex()` on latest commit `5e633dc351346d17ab6b0cb65152d6ad49530768`

DerpyNewbie committed:
Rename/Change method requested on #2

- renamed `GetPosition()` to `GetIndex()`
- changed access specifier of `SetMaterialByType()`
- removed unnecessary comment

DerpyNewbie commented on Source/TicTacToe/TicTacToeBlockGrid.cpp:
```diff
@@ -74,4 +77,128 @@ void ATicTacToeBlockGrid::ChangeTurn()
+	UE_LOG(LogTemp, Verbose, TEXT("Processing turn!"));
+
+	if (CurrentGameState != 0 || executedBlock->BlockType != 0) {
+		return;
```
Should be throwing exception since they were not able to handle turn?

I-K0haku-I replied:
I don't understand why you need an exception. If the turn is not possible then you can't click on it.

Also, you have this check in ```HandleClicked``` in the actual Block already.

DerpyNewbie replied:
Removed if statement on commit `1989be663cc69a2a645061378aae6769fd9c8533`
(i thought it's useful if it throws exception so AI can notice they failed to click ;w;)

I-K0haku-I committed:
add BlockType enum

DerpyNewbie committed:
Add enums for magic numbers

- changed every int to enums
- changed logging little bit
- fixed private access issue

I-K0haku-I commented on Source/TicTacToe/TicTacToeBlock.cpp:
```diff
@@ -62,40 +65,35 @@ void ATicTacToeBlock::OnFingerPressedBlock(ETouchIndex::Type FingerIndex, UPrimi
+	UE_LOG(LogTemp, VeryVerbose, TEXT("TILEARRAYPOS: %d, BLOCKTYPE: %s, INMATCH: %s, WINSTATE: %s"), OwningGrid->GetIndex(this), TOSTR(BlockType), TOSTR(OwningGrid->IsInMatch), TOSTR(OwningGrid->WinState));
+	// Do not highlight if the block has already been activated or if the current game state has not set to 0.
+	if (BlockType != EBlockType::None || !OwningGrid->IsInMatch)
 	{
```
Could make a ```CanClick``` method since this appears twice now. Would be more clear what happens too.

DerpyNewbie replied:
Added `IsClickable()` on commit `e09f9c72704772da6563dc98164db426c90aa862`

I-K0haku-I commented on Source/TicTacToe/TicTacToeBlockGrid.cpp:
```diff
@@ -64,14 +75,141 @@ void ATicTacToeBlockGrid::AddScore()
+	// Set type / material
+	executedBlock->SetType(CurrentType);
+
+	WinState = CheckWin(executedBlock);
```
Rename to ```CheckWinState``` if you want to keep it like this.

DerpyNewbie replied:
Renamed to `CheckWinState()` on commit `1989be663cc69a2a645061378aae6769fd9c8533`

DerpyNewbie committed:
Add logging category

- fixed WinState not resetting on ResetGame()
- added logging category
- removed OnPreNextTurn()
- changed some comments

DerpyNewbie committed:
Add IsClickable()

- added `IsClickable()`
- changed not clickable checks to `!IsClickable()`

I-K0haku-I commented on Source/TicTacToe/TicTacToeBlockGrid.cpp:
```diff
@@ -126,78 +125,80 @@ int ATicTacToeBlockGrid::CheckWin(ATicTacToeBlock* checkBlock)
+	IsInMatch = false;
 
-	//TODO: Implement game end
 	// how do i implement showing reset button? :thinking:
```
Watch Tutorial for UI? 🤔
